tools/mage/end-to-end.go

```diff
@@ -0,0 +1,154 @@
+}
+
+func (EndToEnd) prepareDB() error {
+	isCI := os.Getenv("CI") == "true"
+	dumpExists := pathExists(filepath.Join(".cache", "sqldump.sql"))
+	ok, err := target.Dir(
+		filepath.Join(".cache", "/sqldump.sql"),
+		filepath.Join("pkg", "identityserver", "store"),
+	)
+	if err != nil {
+		return targetError(err)
+	}
+	if !ok || (isCI && dumpExists) {
+		mg.SerialDeps(EndToEnd.DBRestore)
+		return nil
+	}
+	os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+	mg.SerialDeps(Dev.DBErase, Dev.DBStart, Dev.InitStack, EndToEnd.DBDump)
+	return nil
+}
+
+func (EndToEnd) prepareBuild() {
+	mg.SerialDeps(Js.Deps, Js.Build)
+}
+
+// Prepare prepares the server for running end to end tests.
+func (EndToEnd) Prepare() {
+	mg.Deps(EndToEnd.prepareDB, EndToEnd.prepareBuild)
+}
+
+// DBDump performs a database dump to the .cache folder.
+func (EndToEnd) DBDump() error {
+	if mg.Verbose() {
+		fmt.Printf("Execute database dump\n")
+	}
+	_ = os.MkdirAll(".cache", 0755)
+	return execDockerComposeWithOutput(filepath.Join(".cache", "sqldump.sql"), "exec", "-T", "cockroach", "./cockroach", "dump", testDatabaseName, "--insecure")
+}
+
+// DBRestore restores the database using a previously generated dump.
+func (EndToEnd) DBRestore() error {
+	mg.Deps(Js.Deps, Dev.DBStart)
+	if mg.Verbose() {
+		fmt.Printf("Restore database from dump")
+	}
+	return sh.Run("node", filepath.Join("tools", "mage", "scripts", "restore-db-dump.js"), "--db", testDatabaseName)
+}
+
+// Cypress runs the Cypress end-to-end tests.
+func (e EndToEnd) Cypress() error {
+	mg.Deps(Js.Deps, EndToEnd.WaitUntilReady)
+	if mg.Verbose() {
+		fmt.Println("Running Cypress E2E Tests")
+	}
+	return e.runCypress("run")
+}
+
+// CypressOpen runs the Cypress end-to-end tests in interactive mode.
+func (e EndToEnd) CypressOpen() error {
+	mg.Deps(Js.Deps)
+	if mg.Verbose() {
+		fmt.Println("Running Cypress in interactive mode")
+	}
+	return e.runCypress("open")
+}
+
+// StartTestStack starts TTS in end-to-end test configuration.
+func (EndToEnd) StartTestStack() error {
+	mg.Deps(Js.Build)
+	if mg.Verbose() {
+		fmt.Println("Starting stack in end-to-end test configuration")
+	}
+	os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+	return runGoMute("./cmd/ttn-lw-stack", "start")
+}
+
+// WaitUntilReady waits until the web endpoints become available. For CI use.
+func (EndToEnd) WaitUntilReady() error {
+	if mg.Verbose() {
+		fmt.Println("Waiting for the stack to be ready...")
+	}
+	for i := 0; i < 100; i++ {
+		resp, _ := http.Get(getTestURL() + "/oauth")
+		if resp != nil && resp.StatusCode == 200 {
+			return nil
+		}
+		time.Sleep(1000 * time.Millisecond)
+	}
+	return errors.New("Could not connect to server")
+}
+
+// RunFrontend starts the frontend end-to-end tests from scratch.
+func (EndToEnd) RunFrontend() {
+	if mg.Verbose() {
+		fmt.Println("Running end-to-end frontend based tests")
+	}
+	mg.Deps(EndToEnd.Cypress)
+}
+
+// Run starts the end-to-end tests from scratch.
+func (EndToEnd) Run() {
```

Comment:
This does not do anything, remove

Reply:
I [commented](https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442262959) before on why I used this.

Reply:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r447592674